Kotlin implementation template pattern

abhiram-shaji opened the pull request:
This pull request introduces a Kotlin-based implementation of the Template Method pattern. The primary goal is to demonstrate how this pattern can be utilized in Kotlin to structure code that follows a specific algorithmic sequence, allowing for customization of certain steps without altering the overall workflow.

**NameProcessor Abstract Class:** So, I've crafted an abstract class named NameProcessor. This bad boy defines the template method and the steps that need to be executed. These steps are like abstract methods within the class, meaning that any subclasses have got to provide the nitty-gritty details for each step.

**Concrete Subclass:**  This subclass is where the magic happens, showing how the template method can lead the workflow dance, with a little help from the concrete implementations in the subclass.

**Main Method:** Added a main method that demonstrates the use of the template pattern in action. It's like the stage where our subclass gets to shine, executing the defined algorithm and sticking to the script provided by the template method.

**README.md:** Included a README.md file that provides a comprehensive explanation of the Template Method pattern, the ins and outs of its implementation in this project, and the steps to run the code so you can witness this pattern doing its thing. Voilà!

abhiram-shaji committed:
Add the abstract class with the template method

abhiram-shaji committed:
Add the concrete class implementation

abhiram-shaji committed:
Add main function to demonstrate names database usage

abhiram-shaji committed:
Add comments for better code understanding

abhiram-shaji committed:
Create readme file for template pattern implementation

Dhuds1 commented on patterns/Creational/template/Kotlin/template.kt:
```diff
@@ -0,0 +1,47 @@
+// Abstract class defining the template method and steps
+abstract class NamesTemplate {
+    // Template method to manage names
+    fun manageNames(name: String) {
+        println("Starting to manage the name: $name")
+        addName(name)
+        checkName(name)
+        listNames()
+        println("Finished managing the name: $name\n")
+    }
```
In your example, when adding a name there are a lot of events messages that get triggered after you call `manageNames`. Although I don't think this is necessarily the wrong approach, I feel like it could hinder the experience, as it prints every action that takes place.

I think there should be a better way of doing it, where you only get one success message saying "$name has been successfully added to the database".

Because as of right now a user will get this message per name added to the database.
`Starting to manage the name: $name`,
`Added: $name to the database.`
`Checked: $name exists in the database` or `Checked: $name does not exist in the database`
`Current names in the database: [Array]`
`Finished managing the name: $name`

Dhuds1 reviewed (commented):
Hey! I think that you're code is pretty good, but I think we could clean up the messages slightly! I've refactored your code to reduce the amount of messages that the user receives. Also I'm not sure if you're wanting to only add unique names, but if that's the goal consider adding a conditional to check if a name is already in the DB. Let me know what you think of my refactor suggestions!

```Kotlin
// Abstract class defining the template method and steps
abstract class NamesTemplate {
    // Template method to manage names
    fun manageNames(name: String) {
        println("Starting to manage the name: $name")
        val response = addName(name)
        if (response) {
            println("$name added successfully. Names in DB " + listNames())
        } else {
            println("Failed to add the $name")
        }
        println("Finished managing the name: $name\n")
    }

    // Steps to be implemented by subclasses
    abstract fun addName(name: String): Boolean
    abstract fun checkName(name: String): Boolean
    abstract fun listNames(): MutableList<String>
}

// Concrete class implementing the steps for the names database
class NamesDatabase : NamesTemplate() {
    private val namesList = mutableListOf<String>()

    override fun addName(name: String): Boolean {
        namesList.add(name)
        return checkName(name)
    }

    override fun checkName(name: String): Boolean {
        return namesList.contains(name)
    }

    override fun listNames(): MutableList<String> {
        return namesList.toMutableList()
    }
}
```

manasvi111 commented:
hello @abhiram-shaji,

as of now your code looks pretty good and straightforward as anybody can understand by looking at the code. The only things I would like to point you is to add input validation logic within the 'manageNames' method to ensure that the names being added or checked meet certain criteria. This modification would enhance the robustness of the code by preventing invalid or unexpected input from causing issues in the database.

Overall Good job! Abhi.

abhiram-shaji commented:
Thanks @Dhuds1 and @manasvi111 for the reviews, I will definitely consider these points in account for modification.

abhiram-shaji committed:
Refactor namestemplate and namesdatabase for improved feedback and control

abhiram-shaji committed:
Update code to implement input validation

abhiram-shaji committed:
Merge branch 'nic-dgl104-winter-2024:main' into kotlin-implementation-template-pattern

abhiram-shaji committed:
Update readme file

abhiram-shaji committed:
Merge branch 'kotlin-implementation-template-pattern' of https://github.com/abhiram-shaji/pattern-library into kotlin-implementation-template-pattern

abhiram-shaji commented:
Hi @ash-teach , this work is completed. Thank you !

ash-teach commented on patterns/Creational/template/Kotlin/README.md:
```diff
@@ -0,0 +1,55 @@
+    println("Managing names in the database:")
+    namesDatabase.manageNames("Lara")
+    namesDatabase.manageNames("Katrina")
+}
```
You neglected to close the block code formatting here so the rest of the markdown is getting rendered as code.

ash-teach commented on patterns/Creational/template/Kotlin/README.md:
```diff
@@ -0,0 +1,55 @@
+
+#### Example Usage
+
+- Here’s the scoop in the `main` function example: it showcases our validation logic. The name "Lara" (yep, starting with an uppercase letter) sails through and gets added, while "katrina" (oh no, a lowercase starter) flunks the validation and doesn’t make the cut in the `manageNames` method. Aha, that’s how our validation rolls!
```
Aside from the code block formatting issue referenced above, I think it would be better to integrate the update comments into the overview above.

ash-teach commented on patterns/Creational/template/Kotlin/README.md:
```diff
@@ -0,0 +1,55 @@
+
+    println("Managing names in the database:")
+    namesDatabase.manageNames("Lara")
+    namesDatabase.manageNames("Katrina")
```
Perhaps include some line comments describing the expected output - especially if you are integrating the update content below into the overview.

ash-teach commented on patterns/Creational/template/Kotlin/template.kt:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    override fun listNames(): MutableList<String> {
+        return namesList.toMutableList()
```
`namesList` is already `MutableList` - no need for the cast here.

ash-teach commented on patterns/Creational/template/Kotlin/template.kt:
```diff
@@ -0,0 +1,54 @@
+        return checkName(name)
+    }
+
+    override fun checkName(name: String): Boolean {
```
My expectation when I first read that you were including a `checkName` is that it would be used to check if the database already contains the given name, and would abort on a duplicate. Rather, it seems that `checkName` is being used to validate that the `addName` process worked.

Checking that a name has been added correctly is probably better handled through associated unit tests, and so may be unnecessary here as part of the template pattern implementation. In particular, I think it could be a source of confusion when readers are first encountering this pattern.

ash-teach commented on patterns/Creational/template/Kotlin/template.kt:
```diff
@@ -0,0 +1,54 @@
+    abstract fun checkName(name: String): Boolean
+    abstract fun listNames(): MutableList<String>
+}
+
```
This is a good example overall; however, it doesn't draw on the full power of the template pattern. I would much rather see at least two concrete class implementations to demonstrate the advantages of the template (i.e. swapping out elements to customize subclass/extension behaviour).

ash-teach replied:
However, I wonder if your template might be too narrowly focused? I'm not sure what other concrete classes you might be able to come up with where the `NamesTemplate` and associated abstract methods would be relevant...?

ash-teach commented:
Thanks for this submission, @abhiram-shaji. I've put in a few comments regarding overall formatting and some minor code changes. More critically, I'm not 100% convinced that this example does enough to demonstrate Template. Give my comments some thought, and if you have time to make changes, please go ahead.

abhiram-shaji committed:
Refactor NamesTemplate class and improve documentation

- Fixed markdown formatting to properly close the code block.
- Integrated explanatory comments into the code for better clarity.
- Added inline comments to describe expected outputs in main function.
- Removed redundant casting in listNames method.
- Modified checkName method to check for name existence instead of add operation success.
- Suggested adding more concrete class implementations to showcase template pattern flexibility.
- Assessed and ensured the broad applicability of the NamesTemplate class.

abhiram-shaji reviewed (commented):
- Fixed markdown formatting to properly close the code block.
- Integrated explanatory comments into the code for better clarity.
- Added inline comments to describe expected outputs in main function.
- Removed redundant casting in listNames method.
- Modified checkName method to check for name existence instead of add operation success.
- Suggested adding more concrete class implementations to showcase template pattern flexibility.
- Assessed and ensured the broad applicability of the NamesTemplate class.
